File: javascript/bob/bob.js
/* eslint-disable no-unused-vars */
import { all } from './lib/utilities.js';

const isAllUpper = str => str === str.toUpperCase();

const hasLetter = str => str.match(/[a-z]/i);

const isQuestion = str => str.endsWith('?');

const nonWhiteSpace = str => !str || str.match(/^[\s]+$/);

const isShouting = all(isAllUpper, hasLetter);

const isForcefulQuestion = all(isShouting, isQuestion)

const responses = {
  EMPTY: 'Fine. Be that way!',
  FORCEFUL_QUESTION: 'Calm down, I know what I\'m doing!',
  SHOUTING: 'Whoa, chill out!',
  QUESTION: 'Sure.',
  DEFAULT: 'Whatever.'
}

class Bob {
  hey(message) {
    const trimmed = message.trim();
    const { DEFAULT, EMPTY, FORCEFUL_QUESTION, QUESTION, SHOUTING } = responses
    return (
        nonWhiteSpace(trimmed) ? EMPTY
      : isForcefulQuestion(trimmed) ? FORCEFUL_QUESTION
      : isShouting(trimmed) ? SHOUTING
      : isQuestion(trimmed) ? QUESTION
      : DEFAULT
    );
  }
}

export default Bob;

File: python/bob/bob.py
"""Historical reenactment of the average 80's era U.S. highschooler"""

QUESTION = "Sure."
YELL = "Whoa, chill out!"
YELLED_QUESTION = "Calm down, I know what I'm doing!"
EMPTY = "Fine. Be that way!"
DEFAULT = "Whatever."

def hey(phrase):
    """Historical reenactment of the average 80's era U.S. highschooler"""
    stripped = phrase.strip()
    return (
        YELLED_QUESTION if stripped.isupper() and stripped.endswith('?')
        else QUESTION   if stripped.endswith('?')
        else YELL       if stripped.isupper()
        else EMPTY      if not stripped
        else DEFAULT
    )
